Added form and setup routing, removed pages.inc think thats not needed

thenchev opened the pull request:
First approach on port. noticed late that i can remove also in routing file_entity.file_add_upload_archive. Looking to get some feedback for this now.
1. Patterns is not working correctly same as on d7 will check that out.
2. Removed integration with multiform. Module doesn't seem quite ready yet

Commit:
Added form and setup routing, removed pages.inc think thats not needed

slashrsm commented on file_entity.links.task.yml:
```diff
@@ -20,3 +20,13 @@ entity.file.collection:
+  base_route: entity.file.add_form
+  title: 'File'
+
+file_entity.file_add_archive_form:
```
This should probably be moved to file_entity.links.actions.yml

slashrsm replied:
Yes, you are right. This mimcs D7. Let's keep it this way.

slashrsm commented on file_entity.pages.inc:
```diff
@@ -1,87 +0,0 @@
- * Supports file operations including View, Edit, and Delete.
- */
-
-function file_entity_upload_archive_form($form, &$form_state) {
```
Entry for this still exists in hook_menu(). We should remove it.

slashrsm replied:
Yes, if it is empty now remove it entirely.

slashrsm replied:
WRT hook_menu removal see https://trello.com/c/3nOg4Cpn/37-fe-task-port-code-that-is-left-in-hook-menu-from-d7. 

Summary: Create issue in token issue queue and remove hook_menu from file_entity.

slashrsm commented on file_entity.routing.yml:
```diff
@@ -144,3 +144,14 @@ file_entity.settings:
+  path: /file/add/upload/archive
+  defaults:
+    _form: '\Drupal\file_entity\Form\FileAddArchiveForm'
+    _title: 'Add files'
```
I'd rather call this differently. "Upload files archive", "Upload archive" or something in that direction.

slashrsm commented on src/Form/FileAddArchiveForm.php:
```diff
@@ -0,0 +1,152 @@
+/**
+ * Class FileAddArchiveForm.
+ *
+ * @package Drupal\file_entity\Form
```
Never seen this added to classes. What exactly does this annotation mean?

Berdir replied:
PhpStorm does that by default I think, nothing we use

slashrsm commented on src/Form/FileAddArchiveForm.php:
```diff
@@ -0,0 +1,152 @@
+    $form['upload'] = array(
+      '#type' => 'managed_file',
+      '#title' => $this->t('Upload an archive file'),
+      '#upload_location' => NULL,
```
This is confusing... Why don't we just say "temporary://" here?

slashrsm replied:
Another thing that confuses me here. We are uploading archive to temporary folder. On the other hand we make it managed. This means file can generally be removed at some point but it will stay in DB. It seems this was working like this in D7, but I guess we have great opportunity to re-think it while porting. 

We should either: 
- upload into temporary, leave it unmanaged and clean it up after it was extracted or
- upload it somewhere else and make it managed.

slashrsm commented on src/Form/FileAddArchiveForm.php:
```diff
@@ -0,0 +1,152 @@
+      '#progress_indicator' => 'bar',
+      '#default_value' => $form_state->has('file') ? array($form_state->get('file')->id()) : NULL,
+      '#required' => TRUE,
+      '#description' => 'Files must be less than <strong>' . format_size($validators['file_validate_size'][0]) . '</strong><br> Allowed file types: <strong>' . $validators['file_validate_extensions'][0] . '</strong>',
```
Let's use $this->t() here.

slashrsm commented on src/Form/FileAddArchiveForm.php:
```diff
@@ -0,0 +1,152 @@
+    $form['pattern'] = array(
+      '#type' => 'textfield',
+      '#title' => $this->t('Pattern'),
+      '#description' => $this->t('Only files matching this pattern will be imported. For example, to import all jpg and gif files, the pattern would be <em>*.jpg|*.gif</em>. Use <em>.*</em> to extract all files in the archive.'),
```
_.jpg|_.gif is not correct I believe.

slashrsm commented on src/Form/FileAddArchiveForm.php:
```diff
@@ -0,0 +1,152 @@
+        }
+        $archiver->extract($extract_dir);
+        $pattern = '/' . $form_state->getValue('pattern') . '/';
+        if ($files = file_scan_directory($extract_dir, $pattern)) {
```
It seems that this extracts all files and makes the ones that match pattern managed. Other files are left there. I think we should clean them up. If not we should at lest tell people that everything stays on the server.

slashrsm commented on src/Form/FileAddArchiveForm.php:
```diff
@@ -0,0 +1,152 @@
+   *   An array suitable for passing to file_save_upload() or for a managed_file
+   *   or upload element's '#upload_validators' property.
+   */
+  public function getUploadValidators(array $options = array()) {
```
This seems to be very similar to the same function on \Drupal\file_entity\Form\FileAddForm. Can we merge them and create a trait?

mirodietiker commented on src/Form/FileAddArchiveForm.php:
```diff
@@ -0,0 +1,152 @@
+        drupal_set_message($this->t('Extracted %file and added @count new files.', array('%file' => $archive->getFilename(), '@count' => count($files))));
+      }
+      else {
+        throw new \Exception(t('Cannot extract %file, not a valid archive.', array('%file' => $archive->getFileUri())));
```
FormInterface does not talk about handling exceptions. I think we shoud handle this more gracefully with display error messages.

Commit:
Some cleanup, added remove archive in form, reformat validation we now have a trait

thenchev commented:
1. About links action: left it like this for now. Is there any need to change the position (its the same as in drupal 7)? Maybe we want to make it
   more visible, i can still add it to actions beside add files where we list files.
2. Removed archive item from hook_menu. Since we don't use the whole hook_menu stuff, should i remove it completely.
3. Changed title to upload archive, also changed path so we can go back to file listing. Should i leave it or change it back?
4. Hmm, annotation is from drupal console. Played around a bit. Changed to something more appropriate.
5. So changed this to public:// since i added the checkbox to remove the archive, also made it permanent if we don't check the remove archive checkbox we probably 
   want to keep it.
6. t done.
7. Changed to ._jpg|._gif. Should work now.
8. Probably better to clean them up so added code to remove files that don't match the pattern.

slashrsm commented on src/Form/FileAddArchiveForm.php:
```diff
@@ -0,0 +1,130 @@
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
```
No need to have an override just to call parent function.

slashrsm commented on src/Form/FileAddArchiveForm.php:
```diff
@@ -0,0 +1,130 @@
+        $extract_dir = file_default_scheme() . '://' . pathinfo($archive->getFilename(), PATHINFO_FILENAME);
+        $extract_dir = file_destination($extract_dir, FILE_EXISTS_RENAME);
+        if (!file_prepare_directory($extract_dir, FILE_MODIFY_PERMISSIONS | FILE_CREATE_DIRECTORY)) {
+          throw new \Exception(t('Unable to prepar, e directory %dir for extraction.', array('%dir' => $extract_dir)));
```
This seems like a typo.

slashrsm commented on src/Form/FileAddArchiveForm.php:
```diff
@@ -0,0 +1,130 @@
+        }
+        drupal_set_message($this->t('Extracted %file and added @count new files.', array('%file' => $archive->getFilename(), '@count' => count($files))));
+        if ($form_state->getValue('remove_archive')) {
+          drupal_set_message($this->t('Removed archive.'));
```
Maybe something like "Archive %name was removed from the system."

slashrsm commented:
Cool. Looks OK. Just few cosmetic comments. It also seems that we have test fails. Would also be great to have test coverage for this. :)

Commit:
Addressed feedback, some cleanup

thenchev commented:
Here is some small cleanup... next i will look into the tests...

Commit:
Fix for failing tests

Commit:
Archive upload test

Commit:
attempt at fixing the failing test

Commit:
Fix test, failing test doesn't throw exception anymore
